### PR TITLE
Explicitly halt MIDI playback during MPU Reset and when passed through UART-mode

### PR DESCRIPTION
**Pull request by kcgen**
Some MIDI devices don't stop their MIDI playback when receiving a reset in UART mode.

For example, when International Soccer Challenge (1990) passed back-to-back `0xFF`'s to the MT-32 via UART mode, this results in a hanging note. 

So this PR explicitly halts MIDI playback (by sending "note-off" messages and controller reset requests, for  all MIDI channels), during MPU resets and also when the reset command (0xFF) is passed to the MIDI device via the MPU's UART interface. 

Ref: https://steinberg.help/cubase_pro_artist/v9/en/cubase_nuendo/topics/recording/recording_reset_function_c.html

> "The Reset function sends out note-off messages and resets controllers on all MIDI channels. This is sometimes necessary if you experience hanging notes, constant vibrato, etc. when punching in and out on MIDI recordings with pitchbend or controller data."

### CODE REVIEW

**Comment by kcgen** [comment]
Thanks to @Python-Exoproject for reporting this.

File: src/midi/midi.cpp [hunk 1]
```diff
@@ -203,6 +203,30 @@ void MIDI_RawOutByte(uint8_t data)
...
+	// from the first to last channel
+	for (uint8_t channel = 0xb0; channel <= 0xbf; ++channel) {
+		message[0] = channel;
+
+		message[1] = all_notes_off;
+		PlayMsg(message);
+
+		message[1] = all_controllers_off;
+		PlayMsg(message);
+	}
```
Comment (johnnovak) [inline src/midi/midi.cpp:221]: Won't this send random garbage in the third byte? Probably doesn't matter, but I'm slightly uncomfortable with it 😄 Better to explicitly set to zero, as suggested [here](https://users.cs.cf.ac.uk/dave/Multimedia/node158.html)?

File: src/midi/midi.cpp [hunk 2]
```diff
@@ -203,6 +203,30 @@ void MIDI_RawOutByte(uint8_t data)
...
 }
 
+void MidiHandler::HaltSequence()
+{
+	uint8_t message[3] = {}; // see MIDI_evt_len for length lookup-table
+	constexpr uint8_t all_notes_off = 0x7b;
+	constexpr uint8_t all_controllers_off = 0x79;
+
+	// from the first to last channel
+	for (uint8_t channel = 0xb0; channel <= 0xbf; ++channel) {
```
Comment (johnnovak) [inline src/midi/midi.cpp:213]: Define `0xb0` as `STATUS_CONTROL_CHANGE` or something similar somewhere?

**Review by johnnovak — APPROVED** [review]
Looks fine, some minor comments only

File: src/midi/midi.cpp (hunk 1 above)
Comment (kcgen) [reply src/midi/midi.cpp:221]: Line 208 initializes the array:
```
uint8_t message[3] = {};
```

Which assigns zero to all elements (so `[message[2]` is zero).

**Comment by kcgen** [comment]
Thanks for the review, @johnnovak.
Comments addressed; tests are running as expected.
Merging.

File: src/midi/midi.cpp (hunk 1 above)
Comment (johnnovak) [reply src/midi/midi.cpp:222]: Right, good to know. My C++ is quite rusty, indeed!

Comment (kcgen) [reply src/midi/midi.cpp:222]: I'd rather any of us be be rusty on details but sharp on intent (avoiding garbage/uninitialized values), and appreciate a second set of eyes and words of caution. Thanks, @johnnovak 👍

**Pull request merged**